File: apps/app-frontend/src/helpers/curseforge.ts
import type { ExternalContentSource } from '@modrinth/ui'
import { getVersion } from '@tauri-apps/api/app'
import { fetch as tauriFetch } from '@tauri-apps/plugin-http'
import CurseForgeClient, {
	type File as CurseForgeFile,
	FileRelationType,
	HashAlgo,
	type Mod,
	ModLoaderType,
} from 'curseforge-js'

import { config } from '@/config'

import { install_external_file } from './instance'
import type { ContentFileProjectType, InstanceLoader } from './types'

/** Where the Browse page searches for content. */
export type ContentSource = 'modrinth' | 'curseforge'

export const MINECRAFT_GAME_ID = 432

export type CurseForgeContentType = 'mod' | 'resourcepack' | 'shader' | 'datapack'

/** CurseForge "class" ids for each supported content type. */
export const CURSEFORGE_CLASS_IDS: Record<CurseForgeContentType, number> = {
	mod: 6,
	resourcepack: 12,
	shader: 6552,
	datapack: 6945,
}

const INSTALL_PROJECT_TYPES: Record<CurseForgeContentType, ContentFileProjectType> = {
	mod: 'mod',
	resourcepack: 'resourcepack',
	shader: 'shaderpack',
	datapack: 'datapack',
}

const LOADER_TYPES: Partial<Record<InstanceLoader, ModLoaderType[]>> = {
	forge: [ModLoaderType.Forge],
	neoforge: [ModLoaderType.NeoForge],
	fabric: [ModLoaderType.Fabric],
	quilt: [ModLoaderType.Quilt, ModLoaderType.Fabric],
}

/** Loader names as they appear in a CurseForge file's `gameVersions`. */
const LOADER_TAGS: Partial<Record<ModLoaderType, string>> = {
	[ModLoaderType.Forge]: 'forge',
	[ModLoaderType.NeoForge]: 'neoforge',
	[ModLoaderType.Fabric]: 'fabric',
	[ModLoaderType.Quilt]: 'quilt',
}

export function isCurseForgeContentType(type: string): type is CurseForgeContentType {
	return type in CURSEFORGE_CLASS_IDS
}

export function hasCurseForgeApiKey(): boolean {
	return !!config.curseforgeApiKey
}

let clientPromise: Promise<CurseForgeClient> | null = null

export function getCurseForgeClient(): Promise<CurseForgeClient> {
	clientPromise ??= getVersion().then(
		(version) =>
			new CurseForgeClient({
				apiKey: config.curseforgeApiKey,
				userAgent: `refract/${version}`,
				fetch: tauriFetch as typeof globalThis.fetch,
			}),
	)
	return clientPromise
}

/** Mod loaders to filter by for an instance, or none when the content type doesn't depend on a loader. */
export function getLoaderTypes(
	contentType: CurseForgeContentType,
	loader: InstanceLoader,
): ModLoaderType[] {
	return contentType === 'mod' ? (LOADER_TYPES[loader] ?? []) : []
}

function matchesLoader(file: CurseForgeFile, loaderTypes: ModLoaderType[]): boolean {
	if (loaderTypes.length === 0) return true
	const tags = new Set(file.gameVersions.map((version) => version.toLowerCase()))
	return loaderTypes.some((type) => {
		const tag = LOADER_TAGS[type]
		return tag !== undefined && tags.has(tag)
	})
}

/** Picks the newest file for the game version and loaders, preferring releases over betas and alphas. */
export function pickCompatibleFile(
	files: CurseForgeFile[],
	gameVersion: string,
	loaderTypes: ModLoaderType[],
): CurseForgeFile | null {
	const compatible = files
		.filter(
			(file) =>
				file.isAvailable &&
				!file.isServerPack &&
				file.gameVersions.includes(gameVersion) &&
				matchesLoader(file, loaderTypes),
		)
		.sort((a, b) => {
			if (a.releaseType !== b.releaseType) return a.releaseType - b.releaseType
			return Date.parse(b.fileDate) - Date.parse(a.fileDate)
		})
	return compatible[0] ?? null
}

export async function findCompatibleFile(
	modId: number,
	gameVersion: string,
	loaderTypes: ModLoaderType[],
): Promise<CurseForgeFile | null> {
	const client = await getCurseForgeClient()
	const { data } = await client.files.list(modId, { gameVersion, pageSize: 50 })
	return pickCompatibleFile(data, gameVersion, loaderTypes)
}

export function getSha1(file: CurseForgeFile): string | null {
	return file.hashes.find((hash) => hash.algo === HashAlgo.Sha1)?.value.toLowerCase() ?? null
}

export function toExternalSource(mod: Mod, file: CurseForgeFile): ExternalContentSource {
	const author = mod.authors[0]
	return {
		platform: 'curseforge',
		project_id: String(mod.id),
		file_id: String(file.id),
		project_slug: mod.slug,
		project_title: mod.name,
		project_icon_url: mod.logo?.thumbnailUrl || mod.logo?.url || null,
		project_url: mod.links?.websiteUrl || null,
		author_name: author?.name ?? null,
		author_url: author?.url ?? null,
		file_display_name: file.displayName,
		file_date: file.fileDate,
	}
}

export type CurseForgeInstallTarget = {
	instanceId: string
	gameVersion: string
	loader: InstanceLoader
	/** CurseForge project ids already in the instance; dependencies in this set are skipped. */
	installedProjectIds: Set<string>
}

export type CurseForgeInstallResult = {
	installed: Mod[]
	/** Projects whose authors don't allow downloads from third-party apps, to be downloaded manually. */
	restricted: Mod[]
}

export class NoCompatibleFileError extends Error {
	constructor(readonly mod: Mod) {
		super(`${mod.name} has no files for this Minecraft version and loader`)
	}
}

/** Installs a CurseForge project into an instance along with its required dependencies. */
export async function installCurseForgeProject(
	mod: Mod,
	contentType: CurseForgeContentType,
	target: CurseForgeInstallTarget,
): Promise<CurseForgeInstallResult> {
	const client = await getCurseForgeClient()
	const loaderTypes = getLoaderTypes(contentType, target.loader)
	const result: CurseForgeInstallResult = { installed: [], restricted: [] }
	const visited = new Set(target.installedProjectIds)

	async function install(project: Mod, isDependency: boolean) {
		if (visited.has(String(project.id))) return
		visited.add(String(project.id))

		const file = await findCompatibleFile(project.id, target.gameVersion, loaderTypes)
		if (!file) {
			if (isDependency) return
			throw new NoCompatibleFileError(project)
		}

		const requiredIds = file.dependencies
			.filter((dependency) => dependency.relationType === FileRelationType.RequiredDependency)
			.map((dependency) => dependency.modId)
			.filter((id) => !visited.has(String(id)))
		if (requiredIds.length > 0) {
			const dependencies = await client.mods.getMods({ modIds: requiredIds })
			for (const dependency of dependencies) {
				await install(dependency, true)
			}
		}

		const sha1 = getSha1(file)
		if (!file.downloadUrl || !sha1 || project.allowModDistribution === false) {
			result.restricted.push(project)
			return
		}

		await install_external_file(target.instanceId, {
			url: file.downloadUrl,
			file_name: file.fileName,
			sha1,
			project_type: INSTALL_PROJECT_TYPES[contentType],
			source: toExternalSource(project, file),
		})
		result.installed.push(project)
	}

	await install(mod, false)
	return result
}

File: packages/app-lib/src/util/curseforge.rs
//! A minimal client for the CurseForge Core API, for work that runs in the background such as
//! installing modpacks.

use crate::state::{ExternalContentSource, ExternalPlatform, ProjectType};
use crate::util::fetch::{FetchSemaphore, fetch_advanced};
use reqwest::Method;
use serde::Deserialize;
use serde::de::DeserializeOwned;
use sqlx::SqlitePool;

const API_URL: &str = "https://api.curseforge.com/v1/";

#[derive(Deserialize)]
struct Response<T> {
    data: T,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CurseForgeMod {
    pub id: u32,
    pub name: String,
    pub slug: String,
    pub class_id: Option<u32>,
    #[serde(default)]
    pub authors: Vec<CurseForgeAuthor>,
    pub logo: Option<CurseForgeAsset>,
    pub links: Option<CurseForgeLinks>,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CurseForgeAuthor {
    pub id: u32,
    pub name: String,
    pub url: Option<String>,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CurseForgeAsset {
    pub thumbnail_url: Option<String>,
    pub url: Option<String>,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CurseForgeLinks {
    pub website_url: Option<String>,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CurseForgeFile {
    pub id: u32,
    pub display_name: String,
    pub file_name: String,
    pub file_date: String,
    pub file_length: u64,
    /// Missing when the author doesn't allow downloads from third-party apps.
    pub download_url: Option<String>,
    #[serde(default)]
    pub hashes: Vec<CurseForgeFileHash>,
}

#[derive(Clone, Debug, Deserialize)]
pub(crate) struct CurseForgeFileHash {
    pub value: String,
    /// 1 is SHA-1, 2 is MD5.
    pub algo: u8,
}

impl CurseForgeFile {
    pub fn sha1(&self) -> Option<String> {
        self.hashes
            .iter()
            .find(|hash| hash.algo == 1)
            .map(|hash| hash.value.to_lowercase())
    }
}

impl CurseForgeMod {
    pub fn icon_url(&self) -> Option<String> {
        self.logo.as_ref().and_then(|logo| {
            logo.thumbnail_url.clone().or_else(|| logo.url.clone())
        })
    }

    /// The content type of the project, for the classes that can be installed into an instance.
    pub fn project_type(&self) -> Option<ProjectType> {
        match self.class_id? {
            6 => Some(ProjectType::Mod),
            12 => Some(ProjectType::ResourcePack),
            6552 => Some(ProjectType::ShaderPack),
            6945 => Some(ProjectType::DataPack),
            _ => None,
        }
    }

    /// Where `file` of this project came from, as recorded for installed content.
    pub fn external_source(
        &self,
        file: &CurseForgeFile,
    ) -> ExternalContentSource {
        let author = self.authors.first();
        ExternalContentSource {
            platform: ExternalPlatform::CurseForge,
            project_id: self.id.to_string(),
            file_id: file.id.to_string(),
            project_slug: Some(self.slug.clone()),
            project_title: self.name.clone(),
            project_icon_url: self.icon_url(),
            project_url: self
                .links
                .as_ref()
                .and_then(|links| links.website_url.clone()),
            author_id: author.map(|author| author.id.to_string()),
            author_name: author.map(|author| author.name.clone()),
            author_url: author.and_then(|author| author.url.clone()),
            file_display_name: Some(file.display_name.clone()),
            file_date: Some(file.file_date.clone()),
        }
    }
}

fn api_key() -> crate::Result<&'static str> {
    option_env!("CURSEFORGE_API_KEY")
        .filter(|key| !key.is_empty())
        .ok_or_else(|| {
            crate::ErrorKind::InputError(
                "CurseForge isn't available because this build has no CurseForge API key".to_string(),
            )
            .into()
        })
}

async fn request<T: DeserializeOwned>(
    method: Method,
    path: &str,
    body: Option<serde_json::Value>,
    semaphore: &FetchSemaphore,
    pool: &SqlitePool,
) -> crate::Result<T> {
    let bytes = fetch_advanced(
        method,
        &format!("{API_URL}{path}"),
        None,
        body,
        Some(("x-api-key", api_key()?)),
        None,
        None,
        None,
        semaphore,
        pool,
    )
    .await?;
    Ok(serde_json::from_slice::<Response<T>>(&bytes)?.data)
}

pub(crate) async fn get_file(
    mod_id: u32,
    file_id: u32,
    semaphore: &FetchSemaphore,
    pool: &SqlitePool,
) -> crate::Result<CurseForgeFile> {
    request(
        Method::GET,
        &format!("mods/{mod_id}/files/{file_id}"),
        None,
        semaphore,
        pool,
    )
    .await
}

/// Returns the files with the given ids. Files that no longer exist are left out.
pub(crate) async fn get_files(
    file_ids: &[u32],
    semaphore: &FetchSemaphore,
    pool: &SqlitePool,
) -> crate::Result<Vec<CurseForgeFile>> {
    if file_ids.is_empty() {
        return Ok(Vec::new());
    }
    request(
        Method::POST,
        "mods/files",
        Some(serde_json::json!({ "fileIds": file_ids })),
        semaphore,
        pool,
    )
    .await
}

/// Returns the projects with the given ids. Projects that no longer exist are left out.
pub(crate) async fn get_mods(
    mod_ids: &[u32],
    semaphore: &FetchSemaphore,
    pool: &SqlitePool,
) -> crate::Result<Vec<CurseForgeMod>> {
    if mod_ids.is_empty() {
        return Ok(Vec::new());
    }
    request(
        Method::POST,
        "mods",
        Some(serde_json::json!({ "modIds": mod_ids })),
        semaphore,
        pool,
    )
    .await
}
